### PR TITLE
Adds simple case to help motivate @extract_fields

### PR DESCRIPTION
**Pull request by skrawcz**
If someone wanted to use Hamilton to model a modeling dataflow,
they would struggle. Need a new decorator to handle extracting
outputs from functions that return multiple things and aren't
a dataframe.

## Additions

- hello world showing how to create a model using hamilton as a dataflow language.
- adds decorator `@extract_fields`

## Testing

1. Unit tests & this hello world example.

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows the standards laid out in the [TODO link to standards]()
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Python

- [x] python 3.6
- [x] python 3.7

### CODE REVIEW

File: examples/model_examples/scikit-learn/model_logic.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+def flattened_digits(digit_data: np.ndarray) -> np.ndarray:
+    return digit_data.reshape((len(digit_data), -1))
+
+
+@function_modifiers.config.when(clf='svm')
+def prefit_clf__svm(gamma: float = 0.001) -> svm.SVC:
+    return svm.SVC(gamma=gamma)
+
+
+@function_modifiers.extract_columns(*['X_train', 'X_test', 'y_train', 'y_test'])
```
Comment (elijahbenizzy) [inline examples/model_examples/scikit-learn/model_logic.py:20]: Nit -- `*` is a little silly here :)

Comment (elijahbenizzy) [reply examples/model_examples/scikit-learn/model_logic.py:20]: The point is that this would currently break, right?

**Review by elijahbenizzy — COMMENTED** [review]
Like it -- I think there are a lot of `extract_` we might want. Some options:

1. Have a single extract decorator that can assign polymorphically to different types that we can add extractable types easily
2. Have an extract decorator for each type, share some abstraction (perhaps (1)?)
3. Be opinionated about the things we want to extract -- only allow dfs, typeddicts, matrices (maybe...)

File: examples/model_examples/scikit-learn/model_logic.py (hunk 1 above)
Comment (skrawcz) [reply examples/model_examples/scikit-learn/model_logic.py]: changed this.

**Comment by skrawcz** [comment]
@elijahbenizzy I punted on TypedDict, in favor of something slightly simpler. 
We just need to add unit tests...

File: examples/model_examples/scikit-learn/run.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+from hamilton import driver
+from hamilton import base
+
+import my_train_evaluate_logic
+import digit_loader
+import iris_loader
+
+
+def get_data_loader(data_set: str):
```
Comment (elijahbenizzy) [inline examples/model_examples/scikit-learn/run.py:13]: Pros/cons of using this versus `@config.when` on the dataset? These were both recommended approaches I had in the bets practices btw.

File: examples/model_examples/scikit-learn/run.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    """Returns the module to load that will procur data -- the data loaders all have to define the same functions."""
+    if data_set == 'iris':
+        return iris_loader
+    elif data_set == 'digits':
+        return digit_loader
+    else:
+        raise ValueError(f'Unknown data_name {data_set}.')
+
+
+def get_model_config(model_type: str) -> dict:
```
Comment (elijahbenizzy) [inline examples/model_examples/scikit-learn/run.py:23]: Might be nice to put these in json files -- cold be a simpler way to demonstrate storing/managing configs.

File: examples/model_examples/scikit-learn/run.py (hunk 2 above)
Comment (elijahbenizzy) [reply examples/model_examples/scikit-learn/run.py:13]: Yeah, thinking that this is probably best -- we should think through a way to codify the interfaces here.

File: hamilton/function_modifiers.py [hunk 4]
```diff
@@ -190,6 +192,96 @@ def extractor_fn(column_to_extract: str = column, **kwargs) -> pd.Series:  # avo
         return output_nodes
 
 
+
+class extract_fields(NodeExpander):
+    """Extracts fields from a dictionary of output."""
+
+    def __init__(self, fields: dict, fill_with: Any = None):
```
Comment (elijahbenizzy) [inline hamilton/function_modifiers.py:199]: Fields should be `Dict[str, Type[Type]]` -- note that the keys have to be strings, and the values have to be keys.

File: hamilton/function_modifiers.py [hunk 5]
```diff
@@ -190,6 +192,96 @@ def extractor_fn(column_to_extract: str = column, **kwargs) -> pd.Series:  # avo
...
+        - 1 function that outputs the original dict
+
+        :param fields: Fields to extract. A dict of 'field_name' -> 'field_type'.
+        :param fill_with: If you want to extract a field that doesn't exist, do you want to fill it with a default value?
+        Or do you want to error out? Leave empty/None to error out, set fill_value to dynamically create a field value.
+        """
+        if not fields:
+            raise InvalidDecoratorException('Error an empty dict, or no dict, passed to extract_fields decorator.')
+        elif not isinstance(fields, dict):
+            raise InvalidDecoratorException(f'Error, please pass in a dict or typed dict, not {type(fields)}')
```
Comment (elijahbenizzy) [inline hamilton/function_modifiers.py:211]: Why would fields be a typeddict? If so the types would be `Dict[str, Type[Type]]` or something like that. Again, see point above.

File: hamilton/function_modifiers.py (hunk 4 above)
Comment (elijahbenizzy) [reply hamilton/function_modifiers.py:199]: If we choose to use a TypedDict, `fields` can be optional, validated, and type-safe. So a lot cleaner code but also trickier to read...

File: hamilton/function_modifiers.py [hunk 6]
```diff
@@ -190,6 +192,96 @@ def extractor_fn(column_to_extract: str = column, **kwargs) -> pd.Series:  # avo
...
+        self.fields = fields
+        self.fill_with = fill_with
+
+    def validate(self, fn: Callable):
+        """A function is invalid if it is not annotated with a dict or typing.Dict return type.
+
+        :param fn: Function to validate.
+        :raises: InvalidDecoratorException If the function is not annotated with a dict or typing.Dict type as output.
+        """
+        output_type = inspect.signature(fn).return_annotation
```
Comment (elijahbenizzy) [inline hamilton/function_modifiers.py:232]: Let's be opinionated and force it to be a `Dict[str, Any]`. Again we can use the type-hint if its uniform...

File: hamilton/function_modifiers.py [hunk 7]
```diff
@@ -190,6 +192,96 @@ def extractor_fn(column_to_extract: str = column, **kwargs) -> pd.Series:  # avo
...
+        if typing_inspect.is_generic_type(output_type):
+            base = typing_inspect.get_origin(output_type)
+            if base == dict or base == typing.Dict:
```
Comment (elijahbenizzy) [inline hamilton/function_modifiers.py:235]: Why would `base==dict` if its generic?

**Review by elijahbenizzy — COMMENTED** [review]
Some thoughts 
We have three options for the API:
- Allow `dict`, force specification of types in decorator
- Allow `Dict[str, TYPE]`, force everything to be the same type, just specify names in decorator
- Allow TypedDict, specify everything in decorator

This chooses (1) -- let's think through the API a bit? IMO (2) is the most readable but its also limiting. Do we want to support dicts with varied value-types?

File: hamilton/function_modifiers.py (hunk 4 above)
Comment (skrawcz) [reply hamilton/function_modifiers.py:199]: yep, hence punting on TypedDict support.

File: hamilton/function_modifiers.py (hunk 5 above)
Comment (skrawcz) [reply hamilton/function_modifiers.py:211]: I have to change that string.

File: hamilton/function_modifiers.py (hunk 6 above)
Comment (skrawcz) [reply hamilton/function_modifiers.py:232]: eh. I don't see the value of that code right now.

File: hamilton/function_modifiers.py (hunk 7 above)
Comment (skrawcz) [reply hamilton/function_modifiers.py:235]: good point -- should be able to test for it and remove it if so -- I just forget what type it returns.

Comment (skrawcz) [reply hamilton/function_modifiers.py]: 3.6 returns `typing.Dict` 3.7+ returns dict.

**Review by elijahbenizzy — APPROVED** [review]

**Pull request merged**